### PR TITLE
GPL version of Data Masking for MySQL v8.0

### PR DESCRIPTION
**Pull request by fmbiete**
This is a GPL implementation of Oracle MySQL Enterprise Data Masking plugin. 
It follows the API described in the documentation.

### CODE REVIEW

**Comment by fmbiete** [comment]
Travis jobs are failing compiling other features. As far as I can see, clang and gcc 4.8 in Linux don't include the required support to compile MySQL server.

**Comment by bajrang0789** [comment]
are there any progress on this one, to support the `data_masking.so` plugin on Official Percona Repo. (for MySQL 5.7 as well) ?

**Comment by fmbiete** [comment]
> are there any progress on this one, to support the `data_masking.so` plugin on Official Percona Repo. (for MySQL 5.7 as well) ?

Waiting for Percona review, after that I will add 5.7 support

**Comment by laurynas-biveinis** [comment]
Thank you for this pull request! It is still in our queue and we will process it, unfortunately at the moment I cannot give an ETA

File: plugin/data_masking/src/udf/udf_gen_blacklist.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_gen_blacklist.cc]: Yes, otherwise how do you prevent a concurrent removal of the dictionary?
Other simple idea to improve performance is to use rwlocks

File: plugin/data_masking/include/udf/udf_gen_range.h
Comment (satya-bodapati) [inline plugin/data_masking/include/udf/udf_gen_range.h:21]: I would suggest to aggegrate all the  generate functions (udf_*gen_*()) into one .h and .cc file.

File: plugin/data_masking/src/udf/udf_gen_dictionary.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_gen_dictionary.cc]: Same as above. Answer is yes.
Preferrably use RW locks.

File: plugin/data_masking/src/udf/udf_gen_dictionary.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_gen_dictionary.cc]: How are we sure that result has memory of  "length" bytes allocated?

File: plugin/data_masking/src/udf/udf_gen_rnd_us_phone.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_gen_rnd_us_phone.cc:64]: When is this memory released?

File: plugin/data_masking/src/udf/udf_mask_ssn.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_mask_ssn.cc:76]: is there memory leak here?
Lets say the first mask_ssn() set the initd->ptr to some memory location of 10 bytes (say). The next mask_ssn() would set the initd->ptr to new location without freeing the old one?

Comment (satya-bodapati) [reply plugin/data_masking/src/udf/udf_mask_ssn.cc:76]: Or would the denit() be called for every UDF call()?

File: plugin/data_masking/src/udf/udf_mask_ssn.cc
Comment (satya-bodapati) [inline plugin/data_masking/src/udf/udf_mask_ssn.cc]: memory allocated with new(), do not use free().

Just use  delete initd->ptr

**Review by satya-bodapati — COMMENTED** [review]
Few more requests
1. Add a test with view example as shown here at end https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/data-masking-usage.html

2. Does this work with multi-byte character sets? If so, add some tests with non-ascii characters. If not, add tests to ensure it doesn't crash server.

File: plugin/data_masking/src/udf/udf_gen_rnd_us_phone.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_gen_rnd_us_phone.cc:64]: deinit is automatically called after each call

File: plugin/data_masking/src/udf/udf_mask_ssn.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_mask_ssn.cc:76]: yes, deinit is automatically called after each UDF call to clean up

File: plugin/data_masking/src/udf/udf_mask_ssn.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_mask_ssn.cc]: Thank you, done!

File: plugin/data_masking/src/udf/udf_gen_dictionary.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_gen_dictionary.cc]: Thank you, done!

File: plugin/data_masking/src/udf/udf_gen_blacklist.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_gen_blacklist.cc]: Thank you, done!

File: plugin/data_masking/include/udf/udf_gen_range.h
Comment (fmbiete) [reply plugin/data_masking/include/udf/udf_gen_range.h:21]: It started like that, but it was a huge file. I think it's more easy to maintain and extend if the methods are in different smaller files.

File: plugin/data_masking/src/udf/udf_gen_dictionary.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_gen_dictionary.cc]: Good catch. result has a fixed length. I replaced this with initid->ptr to allow using bigger dictionary terms.

**Comment by satya-bodapati** [comment]
Thank you for addressing the review comments. Your patch looks good. :+1: 

You might have to squash this into single commit. I will check and get back to you.

**Review by satya-bodapati — APPROVED** [review]

File: plugin/data_masking/src/udf/udf_registration.cc
Comment (dutow) [inline plugin/data_masking/src/udf/udf_registration.cc]: This file (and some others) seems to be the same as in the group replication plugin?

Please retain the Oracle copyright for files copied from Oracle plugins, and include Oracle in the plugin authors.

**Review by dutow — CHANGES_REQUESTED** [review]

File: plugin/data_masking/src/udf/udf_registration.cc
Comment (satya-bodapati) [reply plugin/data_masking/src/udf/udf_registration.cc]: May I know why this conversation is marked resolved? Isn't the copyright requirement not applicable anymore?

**Comment by satya-bodapati** [comment]
Request to squash the changes into single commit. After this QA will test the PR

File: plugin/data_masking/src/udf/udf_registration.cc
Comment (fmbiete) [reply plugin/data_masking/src/udf/udf_registration.cc]: Sorry, I forgot to commit-push

File: plugin/data_masking/include/udf/udf_registration.h
Comment (dutow) [inline plugin/data_masking/include/udf/udf_registration.h]: Shouldn't this file have the same copyright as the cc file?

File: plugin/data_masking/src/plugin.cc
Comment (dutow) [inline plugin/data_masking/src/plugin.cc:64]: There's an author field here too

File: plugin/data_masking/include/udf/udf_registration.h
Comment (fmbiete) [reply plugin/data_masking/include/udf/udf_registration.h]: Added Oracle

File: plugin/data_masking/src/plugin.cc
Comment (fmbiete) [reply plugin/data_masking/src/plugin.cc:64]: I'm not sure about this one, wouldn't this that mean that Oracle is the author of this plugin? but added anyway.

Comment (satya-bodapati) [reply plugin/data_masking/src/plugin.cc:64]: I don't think Oracle should be put here.. @dutow what was your concern here?

Comment (satya-bodapati) [reply plugin/data_masking/src/plugin.cc:64]: You can remove Oracle here

**Comment by satya-bodapati** [comment]
Hello @fmbiete 

Just want to update that your patch is under testing now

File: plugin/data_masking/src/plugin.cc
Comment (fmbiete) [reply plugin/data_masking/src/plugin.cc:64]: Removed. Thank you

**Comment by dutow** [comment]
@fmbiete I updated your branch. Please update your local checkout and check it.

Changes:
* Added missing DBUG_RETURN macros
* Fixed the tests so they work when executed from the build directory
* Rerecorded the view mtr result, as it was failing
* Squashed your two commits
* Rebased on the current 8.0 head

**Comment by dutow** [comment]
@fmbiete @satya-bodapati I rebased on 8.0.17 and reformatted with clang-format-8.

I'll merge the code once jenkins completes.

Jenkins job: https://ps80.cd.percona.com/job/percona-server-8.0-param/377/

**Comment by satya-bodapati** [comment]
> @fmbiete @satya-bodapati I rebased on 8.0.17 and reformatted with clang-format-8.
> 
> I'll merge the code once jenkins completes.
> 
> Jenkins job: https://ps80.cd.percona.com/job/percona-server-8.0-param/377/

Thanks @dutow

**Comment by dutow** [comment]
I've fixed another issue I didn't notice locally: the install command for the tests was incorrect and unnecessary.

New jenkins: https://ps80.cd.percona.com/job/percona-server-8.0-param/380/

**Pull request merged**
